![Add a relevant banner image here](path_to_image)

# **Flatiron Data Science Capstone Project 1: US Traffic Accidents**


## Overview



## Business Understanding

The US Department of Transportation (DOT) is concerned about the number of traffic accidents across the US and wants to develop strategies to reduce accidents and improve road safety.

For the Department of Transportation (DOT), understanding and reducing traffic accidents is a critical mission that directly impacts public safety, economic costs, and quality of life across the United States.

This analysis matters from a stakeholder and business perspective:

-Economic Impact: Traffic accidents cost billions annually in medical expenses, property damage, and lost productivity, making even small reductions highly valuable.

-Public Safety: As a leading cause of injury and death, reducing traffic accidents directly fulfills DOT's core mandate to protect citizens.

-Infrastructure Prioritization: Data analysis enables strategic allocation of limited infrastructure improvement budgets to highest-risk areas.

-Policy Development: Accident data informs new safety regulations and provides metrics to evaluate existing programs' effectiveness.

-Stakeholder Accountability: Comprehensive analysis demonstrates evidence-based decision-making to Congress, local governments, and the public.

-Cross-Agency Collaboration: Shared data insights can align accident reduction efforts across DOT, law enforcement, and emergency services.

-Technology Integration: Understanding accident patterns guides how emerging vehicle technologies should be regulated to maximize safety benefits.

This project supports DOT's mission by translating complex accident data into actionable insights. By identifying key patterns and risk factors, it empowers smarter infractstructure investment, better regulation and ultimately safer roads for all.

Project Objectives:
1. Identify accident hotspots: This analysis will determine when and where accidents most frequently occur.  Patterns such as time of day, day of week, season, and geographic location will be examined to determine of there are critical hotspots and time periods that may warrant intervention.
2. Analyze environmental risk factors: This analysis will determine how weather conditions correlate with accident rates. Factors such as visibility, precipitation, temperature and other environmental variables will be examined to assess their impact on driver behavior and road conditions. The goal is to determine if certain weather conditions should trigger early warning notifications to drivers.
3. Compare Urban and Rural Accident Patterns: This analysis will assess accident patterns between urban and rural settings. These different environments likely present distinct challenges and risk factors. Understanding these differences can inform the development of location-specific safety strategies.

By successfully identifying accident hotspots, environmental risk factors, and urban/rural accident patterns, DOT can execute initiatives to address these issues and fulfill its mission to the public.

## Data Understanding

Text here

In [1]:
# from google.colab import drive
# drive.mount('/content/drive')

In [2]:
# Load relevant imports here
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels as stats
import scipy

# column definitions in Onedrive


In [3]:
# load data
# df = pd.read_csv('/content/drive/MyDrive/US_Accidents_March23.csv')

# PC path to the CSV file
df = pd.read_csv(r"C:\Users\jtlee\OneDrive\Documents\Flatiron Schoolwork\DS_11 Capstone\Project 1 US Traffic Accidents\US_Accidents_March23.csv")

# add laptop path to CSV file
# df = pd.readcsv(r'')

In [4]:
print(f"\nShape of US accidents dataset: {df.shape}\n")     # (rows, columns)

df.info()  # DataFrame info


Shape of US accidents dataset: (7728394, 46)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7728394 entries, 0 to 7728393
Data columns (total 46 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   ID                     object 
 1   Source                 object 
 2   Severity               int64  
 3   Start_Time             object 
 4   End_Time               object 
 5   Start_Lat              float64
 6   Start_Lng              float64
 7   End_Lat                float64
 8   End_Lng                float64
 9   Distance(mi)           float64
 10  Description            object 
 11  Street                 object 
 12  City                   object 
 13  County                 object 
 14  State                  object 
 15  Zipcode                object 
 16  Country                object 
 17  Timezone               object 
 18  Airport_Code           object 
 19  Weather_Timestamp      object 
 20  Temperature(F)         float64
 21  Wind_Ch

In [5]:
print(f"Print preview of the dataframe:\n")
print(df.head())

Print preview of the dataframe:

    ID   Source  Severity           Start_Time             End_Time  \
0  A-1  Source2         3  2016-02-08 05:46:00  2016-02-08 11:00:00   
1  A-2  Source2         2  2016-02-08 06:07:59  2016-02-08 06:37:59   
2  A-3  Source2         2  2016-02-08 06:49:27  2016-02-08 07:19:27   
3  A-4  Source2         3  2016-02-08 07:23:34  2016-02-08 07:53:34   
4  A-5  Source2         2  2016-02-08 07:39:07  2016-02-08 08:09:07   

   Start_Lat  Start_Lng  End_Lat  End_Lng  Distance(mi)  ... Roundabout  \
0  39.865147 -84.058723      NaN      NaN          0.01  ...      False   
1  39.928059 -82.831184      NaN      NaN          0.01  ...      False   
2  39.063148 -84.032608      NaN      NaN          0.01  ...      False   
3  39.747753 -84.205582      NaN      NaN          0.01  ...      False   
4  39.627781 -84.188354      NaN      NaN          0.01  ...      False   

  Station   Stop Traffic_Calming Traffic_Signal Turning_Loop Sunrise_Sunset  \
0   False 

In [6]:
print(df.isnull().sum())

ID                             0
Source                         0
Severity                       0
Start_Time                     0
End_Time                       0
Start_Lat                      0
Start_Lng                      0
End_Lat                  3402762
End_Lng                  3402762
Distance(mi)                   0
Description                    5
Street                     10869
City                         253
County                         0
State                          0
Zipcode                     1915
Country                        0
Timezone                    7808
Airport_Code               22635
Weather_Timestamp         120228
Temperature(F)            163853
Wind_Chill(F)            1999019
Humidity(%)               174144
Pressure(in)              140679
Visibility(mi)            177098
Wind_Direction            175206
Wind_Speed(mph)           571233
Precipitation(in)        2203586
Weather_Condition         173459
Amenity                        0
Bump      

# Objective 1 accident hotspot analysis

Will look at patterns such as time of day, day of week, season, and geogrpahic location

Start time can be used to determine time of day, day of week, and season.
Geographic location will be determined using state, county and zip code

In [7]:
# Create new dataframe for objective 1 accident hotspot analysis

df_hotspots = df[['ID', 'Start_Time', 'City', 'County', 'State', 'Zipcode', ]].copy()

print(f"Shape of hotspots df: {df_hotspots.shape}")
print(df_hotspots.info())
print(df_hotspots.isnull().sum())


Shape of hotspots df: (7728394, 6)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7728394 entries, 0 to 7728393
Data columns (total 6 columns):
 #   Column      Dtype 
---  ------      ----- 
 0   ID          object
 1   Start_Time  object
 2   City        object
 3   County      object
 4   State       object
 5   Zipcode     object
dtypes: object(6)
memory usage: 353.8+ MB
None
ID               0
Start_Time       0
City           253
County           0
State            0
Zipcode       1915
dtype: int64


253 missing City and 1,915 zipcode values out of 7.7 million data points is not significant. They will be dropped, as they shouldn't hurt the analysis at all

In [8]:
# check for unique states and their counts
print(f"\nUnique states: {sorted(df['State'].unique())}")
print(f"\nNumber of states included in the dataset: {df['State'].nunique()}")



Unique states: ['AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA', 'IA', 'ID', 'IL', 'IN', 'KS', 'KY', 'LA', 'MA', 'MD', 'ME', 'MI', 'MN', 'MO', 'MS', 'MT', 'NC', 'ND', 'NE', 'NH', 'NJ', 'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX', 'UT', 'VA', 'VT', 'WA', 'WI', 'WV', 'WY']

Number of states included in the dataset: 49


It appears that there are only 49 states included in this US accidents dataset. Upon further examination, DC has been included, and Alaska and Hawaii are not present in the data.

## Data Preparation




# Objective 1 Accident hotpot analysis

In [9]:
# drop rows with missing City and Zipcode values
df_hotspots.dropna(subset=['City', 'Zipcode'], inplace=True)
df_hotspots.shape

(7726228, 6)

In [10]:
# take Start Time column and convert to datetime format
df_hotspots['Start_Time'] = pd.to_datetime(df_hotspots['Start_Time'], format='mixed', errors='coerce')

# create new column for time of day (hour of the day)
df_hotspots['Time_of_day'] = df_hotspots['Start_Time'].dt.hour

# create new column for day of week
# df_hotspots['day_of_week'] = df_hotspots['Start_Time'].dt.dayofweek

# convert day of week to actual day name
df_hotspots['Day_of_week'] = df_hotspots['Start_Time'].dt.day_name()

print(df_hotspots.head())

    ID          Start_Time          City      County State     Zipcode  \
0  A-1 2016-02-08 05:46:00        Dayton  Montgomery    OH       45424   
1  A-2 2016-02-08 06:07:59  Reynoldsburg    Franklin    OH  43068-3402   
2  A-3 2016-02-08 06:49:27  Williamsburg    Clermont    OH       45176   
3  A-4 2016-02-08 07:23:34        Dayton  Montgomery    OH       45417   
4  A-5 2016-02-08 07:39:07        Dayton  Montgomery    OH       45459   

   Time_of_day Day_of_week  
0            5      Monday  
1            6      Monday  
2            6      Monday  
3            7      Monday  
4            7      Monday  


In [11]:
# define a function to get the season based on the month
def get_season(month):
    if month in [12, 1, 2]:
        return 'Winter'
    elif month in [3, 4, 5]:
        return 'Spring'
    elif month in [6, 7, 8]:
        return 'Summer'
    elif month in [9, 10, 11]:
        return 'Fall'
    
df_hotspots['Season'] = df_hotspots['Start_Time'].dt.month.apply(get_season)

print(df_hotspots.head())

    ID          Start_Time          City      County State     Zipcode  \
0  A-1 2016-02-08 05:46:00        Dayton  Montgomery    OH       45424   
1  A-2 2016-02-08 06:07:59  Reynoldsburg    Franklin    OH  43068-3402   
2  A-3 2016-02-08 06:49:27  Williamsburg    Clermont    OH       45176   
3  A-4 2016-02-08 07:23:34        Dayton  Montgomery    OH       45417   
4  A-5 2016-02-08 07:39:07        Dayton  Montgomery    OH       45459   

   Time_of_day Day_of_week  Season  
0            5      Monday  Winter  
1            6      Monday  Winter  
2            6      Monday  Winter  
3            7      Monday  Winter  
4            7      Monday  Winter  


In [12]:
# Check of consistency of spelling in top 50 city and county columns
print(df['City'].value_counts().head(50))
print(df['County'].value_counts().head(50))


City
Miami              186917
Houston            169609
Los Angeles        156491
Charlotte          138652
Dallas             130939
Orlando            109733
Austin              97359
Raleigh             86079
Nashville           72930
Baton Rouge         71588
Atlanta             68186
Sacramento          66264
San Diego           55504
Phoenix             53974
Minneapolis         51488
Richmond            48845
Oklahoma City       46092
Jacksonville        42447
Tucson              39304
Columbia            38178
San Antonio         37961
Greenville          37802
Saint Paul          37383
Seattle             36564
Portland            35399
San Jose            34536
Indianapolis        33219
Chicago             32035
Tampa               31193
Denver              30557
Kansas City         30107
Tulsa               28880
Riverside           27410
New Orleans         27354
Bronx               27297
Rochester           26866
Fort Lauderdale     26205
Detroit             24865
Grand R

In [13]:
# extract just the first 5 digits of the Zipcode column
# This is to ensure that the Zipcode column is in the correct format
df_hotspots['Zipcode'] = df_hotspots['Zipcode'].astype(str).str.extract(r'^(\d{5})')
print(df_hotspots['Zipcode'].value_counts().head(50))
print(df_hotspots['Zipcode'].unique())

Zipcode
27610    12907
33186    11952
32819    11894
91761    11455
70808    10971
28205    10797
91706    10556
28208    10411
27604    10185
37211     9931
92407     9925
28216     9832
33169     9809
70816     9762
92507     9548
70802     9381
28217     9354
90022     9333
33155     9225
29615     9168
90023     9153
29210     8965
32837     8817
33168     8817
28273     8792
27603     8727
37013     8641
70809     8415
91748     8389
27612     8226
33150     8222
37210     8142
28262     8137
75243     8083
32839     8078
37207     8009
27606     7913
78753     7870
27616     7809
90248     7724
29223     7719
91765     7695
27529     7634
90012     7565
33157     7536
92324     7515
28206     7450
32809     7407
27609     7265
33165     7200
Name: count, dtype: int64
['45424' '43068' '45176' ... '69120' '38614' '04231']


In [14]:
print(df_hotspots.head())

    ID          Start_Time          City      County State Zipcode  \
0  A-1 2016-02-08 05:46:00        Dayton  Montgomery    OH   45424   
1  A-2 2016-02-08 06:07:59  Reynoldsburg    Franklin    OH   43068   
2  A-3 2016-02-08 06:49:27  Williamsburg    Clermont    OH   45176   
3  A-4 2016-02-08 07:23:34        Dayton  Montgomery    OH   45417   
4  A-5 2016-02-08 07:39:07        Dayton  Montgomery    OH   45459   

   Time_of_day Day_of_week  Season  
0            5      Monday  Winter  
1            6      Monday  Winter  
2            6      Monday  Winter  
3            7      Monday  Winter  
4            7      Monday  Winter  


In [15]:
def convert_hour_to_period(hour):
    if 0 <= hour < 6:
        return 'Late Night'
    elif 6 <= hour < 12:
        return 'Morning'
    elif 12 <= hour < 17:
        return 'Afternoon'
    elif 17 <= hour < 21:
        return 'Evening'
    else:
        return 'Night'

df_hotspots['Time_period'] = df_hotspots['Time_of_day'].apply(convert_hour_to_period)
print(df_hotspots.head())

    ID          Start_Time          City      County State Zipcode  \
0  A-1 2016-02-08 05:46:00        Dayton  Montgomery    OH   45424   
1  A-2 2016-02-08 06:07:59  Reynoldsburg    Franklin    OH   43068   
2  A-3 2016-02-08 06:49:27  Williamsburg    Clermont    OH   45176   
3  A-4 2016-02-08 07:23:34        Dayton  Montgomery    OH   45417   
4  A-5 2016-02-08 07:39:07        Dayton  Montgomery    OH   45459   

   Time_of_day Day_of_week  Season Time_period  
0            5      Monday  Winter  Late Night  
1            6      Monday  Winter     Morning  
2            6      Monday  Winter     Morning  
3            7      Monday  Winter     Morning  
4            7      Monday  Winter     Morning  


## Analysis

Text here

## Evaluation

### Business Insight/Recommendation 1

### Business Insight/Recommendation 2

### Business Insight/Recommendation 3

### Tableau Dashboard link

## Conclusion and Next Steps
Text here